HiDPI icons for window system icons on Windows and Mac

eirikbakke opened the pull request:
(Update: This PR now includes new window system icons for both Windows and
MacOS. Since this is the same work done twice with small visual variations, I am
combining these in a single PR.)

This pull request introduces vector-drawn Icon implementations for the icons
used in the window system's Windows 8 and Aqua (MacOS) LAFs. Specifically, this
replaces most of the bitmap icons for these two LAFs o.n.swing.tabcontrol and
openide.awt modules.

An abstract class VectorIcon is added in the UI Utilities Module (openide.awt)
as a general-purpose starting point for creating new vector icons. It handles
and documents a number of tricky adjustments that are needed to draw icons that
appear sharp on HiDPI screens. This class can be used to implement HiDPI icons
in other LAFs at a later time, e.g. on Linux.

A small utility, VectorIconTester, was written to preview new icons at multiple
resolutions, as well as to compare them with existing bitmap icons. This is less
polished code that is nevertheless included here, in tabcontrol's test package.
A screenshot of its output, as well as screenshots of the NetBeans IDE before
and after the patch on both Windows and MacOS, are attached to the JIRA
tickets below and attached to this pull request as well.

The relevant JIRA tickets are:
https://issues.apache.org/jira/browse/NETBEANS-1238 (for Windows)
https://issues.apache.org/jira/browse/NETBEANS-1260 (for MacOS)

To test this patch on Windows, use Windows 10 on Java 9 or Java 10 with a HiDPI
screen, and use the workaround in
https://issues.apache.org/jira/browse/NETBEANS-1227 to tell Windows to let
NetBeans manage DPI scaling on its own.

To test this patch on MacOS, use a MacBook Pro.

NetBeans screenshots before and after patch on Windows (be sure to view at 100% resolution):
![netbeans-1238 before patch](https://user-images.githubusercontent.com/886243/45390064-dfb62b80-b5eb-11e8-8d15-3f91739c7eb2.png)
![netbeans-1238 after patch](https://user-images.githubusercontent.com/886243/45390063-dfb62b80-b5eb-11e8-8dcc-45fc4ea08879.png)

NetBeans screenshots before and after patch on MacOS (be sure to view at 100% resolution):
<img width="1680" alt="netbeans-1260 macos retina before patch" src="https://user-images.githubusercontent.com/886243/45652942-45029480-baa4-11e8-8706-4fcde7a98939.png">
<img width="1680" alt="netbeans-1260 macos retina after patch" src="https://user-images.githubusercontent.com/886243/45652950-492eb200-baa4-11e8-8b22-dc7b2b00cb4a.png">

Output of the VectorIconTest utility, showing the old icons next to the new scalable ones:
![vectoricontester output windows and macos hidpi icons](https://user-images.githubusercontent.com/886243/45652928-32885b00-baa4-11e8-8807-0337f81c722d.png)

eirikbakke commented:
Note that this PR introduces a new public class VectorIcon in the openide.awt module, so it should increment that module's API version before being merged.

(I don't think I have permission to add the "API Change" label to the pull request.)

mcdonnell-john commented:
Don't know enough to comment on the API change, but I have ran this PR locally and verified the Icons show as expected and do look a lot better.

Thanks @eirikbakke

darkyellow commented:
Are these icons for specific look and feels only? (Hence why only for Mac and Windows and not Linux)

geertjanw commented:
Seems to be agreement here and no objections and these UI changes make NetBeans better, not worse. Merging.

geertjanw commented:
Before merging, let’s discuss the need for an API version change being needed. Has this been done, should something additional be done?

eirikbakke commented:
I moved VectorIcon into org.openide.util.ui together with ImageUtilities, updated the API versions and added an entry to apichanges.xml. Also tested again on Windows to confirm nothing broke.

eirikbakke commented:
This commit is ready to merge if no one objects to the addition of the "VectorIcon" class to the public API of org.openide.awt module.

eirikbakke commented:
@emilianbold Would you mind having a quick look at the API change here, which is just the addition of a new utility class "VectorIcon"?

I looked at the retina-related work you did in the past. This PR should should not interfere with any of that, as this PR touches the LAF-related icons, and does not change ImageUtilities in any way. It would be great to eventually merge your work as well.

(There are basically two possible approaches to doing HiDPI icons: either via MultiResolutionImage, or via vector-painted custom implementations of the Icon class. I think both approaches should be supported--this PR deals with the vector painting approach, while your previous retina work deals with the MultiResolutionImage approach. The MultiResolutionImage approach currently does not work properly on windows, due to https://bugs.openjdk.java.net/browse/JDK-8212226 , but surely it will work in the long-term.)

eirikbakke commented:
I just tested with Darcula on both Windows and MacOS. Darcula is unaffected, except for the non-LAF specific icons (arrow, toolbar_arrow_horizontal/vertical), which do get applied to Darcula as well. Those will need some more contrast on dark themes; I can take care of that in a separate commit.

Looking at Darcula, its LAF icons seem to be based on the Windows 8 theme. So it should be fairly easy to adapt this work to that plugin. But that would have to be done on the Darcula side. Separate work would also have to be done to vectorize icons on whichever LAF is the default on Linux.

Yes, branding attempts that replace the PNGs but not the LAF identifier will be thwarted by this commit. I think that's OK, except for the contrast issue noted above.

I'll merge this patch and start creating some issues for future HiDPI work.

eirikbakke commented:
Fixed dark theme appearance for the three LAF-independent icons at https://github.com/apache/incubator-netbeans/commit/8f1f57bdc2552541b9947fd12ac8dd74796539f6 . Tested on Darcula on Windows.

eirikbakke commented:
Created an issue https://issues.apache.org/jira/browse/NETBEANS-1583 for the future, once we eventually want to update ImageUtilities to work with MultiResolutionImage instances.
